Managers/MapManager.gd

```diff
@@ -142,30 +146,35 @@ func is_room_event_correct(current_room: RoomBase, map: MapBase = current_map) -
 
 	return true
 
-
-## Picks room type randomly, with each type having a set probability [br]
-## Algorithm taken from this: https://stackoverflow.com/a/1761646
-func pick_room_type() -> EventBase:
-	var _total_proba: int = 0
-	for event: GlobalEnums.EventType in GlobalVar.EVENTS_PROBABILITIES:
-		_total_proba += GlobalVar.EVENTS_PROBABILITIES[event]
-
-	var _rand_number: int = randi_range(0, _total_proba - 1)
-	# This is just for the error message
-	var _rand_number_error: int = _rand_number
-
-	for _rand_type: GlobalEnums.EventType in GlobalVar.EVENTS_PROBABILITIES:
-		if _rand_number < GlobalVar.EVENTS_PROBABILITIES[_rand_type]:
-			return GlobalEnums.choose_event_from_type(_rand_type)
-
-		_rand_number -= GlobalVar.EVENTS_PROBABILITIES[_rand_type]
+## Creates the list that contains all events that we will be picking from to populate the map
+## The number of occurences of each event is defined by the total number of rooms times the probability of the event set in Global_var.gd
+## We then add some extra events (deviation) to avoid being locked, there are rules that could prevent the remaining event types to be picked, resulting in a deadlock
+## While the number of each event is not exactly the expected number, it should be close enough to what we want
+func create_event_list() -> void:
+	var total_nb_rooms: int = 0
+	for nb: int in map_width_array:
+		total_nb_rooms += nb
 	
-	# This shouldn't be reached
-	push_error("Error: The random number _rand_number is bigger than the sum of probabilities _total_proba: %s > %s" % [_rand_number_error, _total_proba])
-	return EventMob.new()
+	for event_type: GlobalEnums.EventType in GlobalVar.EVENTS_PROBABILITIES:
+		var event_type_list: Array[GlobalEnums.EventType]
+		var event_count: int = floor(total_nb_rooms * GlobalVar.EVENTS_PROBABILITIES[event_type]/100)
+		# The floor before boss room is composed of heal rooms
+		#if event_type == GlobalEnums.EventType.Heal:
+			#event_count -= 3
```

Comment:
why is this commented ? it seems to make sense to leave it, no ?

Reply:
ah good catch will remove it

Reply:
(misunderstood the question so I deleted my previous comment)

Ah right because while it makes sense, I saw from running that we'd usually get less heal rooms, since heal rooms are quite restricted by the rules. The expected number is 7 heal rooms, and I got 4 to 7 heal rooms on average, and commenting the lines gave 6 to 9 rooms. So it just kind of worked better by removing them, it gave a 'better' distribution I think

Reply:
oh then if it works better let's keep it like that, you can even remove the commented lines i think